lib/cache/collections.go

```diff
@@ -2299,3 +2302,34 @@ func (integrationsExecutor) getReader(cache *Cache, cacheOK bool) services.Integ
 }
 
 var _ executor[types.Integration, services.IntegrationsGetter] = integrationsExecutor{}
+
+type headlessAuthenticationServiceExecutor struct{}
+
+func (headlessAuthenticationServiceExecutor) getAll(ctx context.Context, cache *Cache, loadSecrets bool) ([]*types.HeadlessAuthentication, error) {
+	return cache.headlessAuthenticationsCache.GetHeadlessAuthentications(ctx)
+}
+
+func (headlessAuthenticationServiceExecutor) upsert(ctx context.Context, cache *Cache, resource *types.HeadlessAuthentication) error {
+	return cache.headlessAuthenticationsCache.UpsertHeadlessAuthentication(ctx, resource)
+}
+
+func (headlessAuthenticationServiceExecutor) deleteAll(ctx context.Context, cache *Cache) error {
+	return cache.headlessAuthenticationsCache.DeleteAllHeadlessAuthentications(ctx)
+}
+
+func (headlessAuthenticationServiceExecutor) delete(ctx context.Context, cache *Cache, resource types.Resource) error {
+	ha, ok := resource.(*types.HeadlessAuthentication)
+	if !ok {
+		return trace.BadParameter("unexpected type %T", resource)
+	}
+	return cache.headlessAuthenticationsCache.DeleteHeadlessAuthentication(ctx, ha.User, resource.GetName())
+}
+
+func (headlessAuthenticationServiceExecutor) isSingleton() bool { return false }
+
+func (headlessAuthenticationServiceExecutor) getReader(_ *Cache, _ bool) any {
+	// headless authentication resources aren't directly used by Cache so there's no associated reader type
+	return nil
+}
```

Comment:
I'd use `struct{}` rather than `any` as the reader type for this, just so it's clear that it's a deliberate noop reader rather than anything dynamic.